File: frontend/src/types/workflow.ts
export const WorkflowNodeTypes = {
  Start: 'start',
  Agent: 'agent',
  Condition: 'condition',
  End: 'end',
  Loop: 'loop',
  Parallel: 'parallel',
  Tool: 'tool',
  Human: 'human',
  SubWorkflow: 'subworkflow',
} as const

export type WorkflowNodeType = typeof WorkflowNodeTypes[keyof typeof WorkflowNodeTypes]

export interface INodeConfig {
  [key: string]: unknown
}

export interface IWorkflowNode {
  id: string
  type: WorkflowNodeType
  label: string
  agentId?: string
  config?: string
  x: number
  y: number
}

export interface IWorkflowEdge {
  id: string
  source: string
  target: string
  sourceHandle?: string
  targetHandle?: string
}

export interface IWorkflow {
  id: string
  name: string
  description: string
  nodes: IWorkflowNode[]
  edges: IWorkflowEdge[]
  inputSchema?: string
  variables?: string
  version: number
  isEnabled: boolean
  sortOrder: number
  createdAt: string
  updatedAt: string
}

export interface CreateWorkflowRequest {
  name: string
  description: string
  nodes: IWorkflowNode[]
  edges: IWorkflowEdge[]
  inputSchema?: string
  variables?: string
  isEnabled: boolean
  sortOrder: number
}

export type UpdateWorkflowRequest = CreateWorkflowRequest

export interface IWorkflowRun {
  id: string
  workflowId: string
  status: 'Pending' | 'Running' | 'Succeeded' | 'Failed' | 'Cancelled'
  input?: string
  output?: string
  startedAt?: string
  completedAt?: string
  chatTopicId?: string
  error?: string
  totalIterations: number
  createdAt: string
}

export interface IWorkflowRunNode {
  id: string
  runId: string
  nodeId: string
  nodeType: string
  status: 'Pending' | 'Running' | 'Succeeded' | 'Failed' | 'Skipped'
  input?: string
  output?: string
  startedAt?: string
  completedAt?: string
  error?: string
  iterations: number
}

export interface IWorkflowRunDetail {
  run: IWorkflowRun
  nodes: IWorkflowRunNode[]
}

export interface IValidationResult {
  valid: boolean
  errors: string[]
}

export interface IRunWorkflowRequest {
  input?: string
}

export interface IWorkflowEvent {
  type: string
  runId?: string
  nodeId?: string
  nodeType?: string
  label?: string
  output?: string
  error?: string
  result?: {
    runId: string
    status: string
    output?: string
    error?: string
  }
}

File: src/Hetu.Core/Entities/Workflow.cs
namespace Hetu.Core.Entities;

/// <summary>
/// 工作流定义。由节点（Nodes JSON）和边（Edges JSON）组成有向图，支持环路。
/// 节点类型：Start / Agent / Condition / End / Loop / Parallel / Tool / Human / SubWorkflow。
/// </summary>
public class Workflow : BaseEntity
{
    public string Name { get; set; } = string.Empty;
    public string Description { get; set; } = string.Empty;

    /// <summary>JSON: NodeDto[] — 节点定义（含 id/type/agentId/config/position 等）</summary>
    public string Nodes { get; set; } = "[]";

    /// <summary>JSON: EdgeDto[] — 边定义（含 source/target/sourceHandle/targetHandle）</summary>
    public string Edges { get; set; } = "[]";

    /// <summary>JSON: 输入参数 schema（Start 节点的输入定义）</summary>
    public string? InputSchema { get; set; }

    /// <summary>JSON: 工作流级变量默认值</summary>
    public string? Variables { get; set; }

    public int Version { get; set; } = 1;
    public bool IsEnabled { get; set; } = true;
    public int SortOrder { get; set; }
}
